Firedoors will now close on mobs Part 3

Shadowtail117 opened the pull request:
Remaking this to squash some merge commits. See #28270 for the previous iteration with all of its comments.

:cl:
tweak: Firedoors will no longer refuse to close if something is on top of their tile.
tweak: Firedoors will now give two seconds of warning before closing if a mob is on its tile when it tries to close.
soundadd: Firedoors will play a warning beep if a mob is on its tile when it tries to close.
tweak: If a mob does not move from a firedoor after two seconds, it will push the mob out of the way. This deals a slight amount of damage due to the force of the push. The firedoor will try to move the mob to the safe side of itself.
/:cl:

MistakeNot4892 commented on code/game/machinery/doors/firedoor.dm:
```diff
@@ -348,12 +346,13 @@
-		return
-
 	latetoggle()
+	visible_message(SPAN_DANGER("The firedoor beeps ominously, get out of the way!"), SPAN_DANGER("You hear some beeping coming from the ceiling."), 3)
```
This will stack if `close()` is called repeatedly and does not check if the door is already closed or has started closing.

Shadowtail117 replied:
From what I understand there are very few if not no circumstances where an already-closed firedoor will try to close again outside of debugging/admin hijinks.

Spookerton replied:
iirc spamming a crowbar on an open one may be able to produce a rapid open and close, but the check is worth having even if not trivial to actually do.

MistakeNot4892 commented on code/game/machinery/doors/firedoor.dm:
```diff
@@ -348,12 +346,13 @@
+	for(var/turf/turf in locs)
+		for(var/mob/living/M in turf)
+			if(M.airlock_crush(FIREDOOR_CRUSH_DAMAGE))
+				take_damage(FIREDOOR_CRUSH_DAMAGE)
```
`airlock_crush()` already applies damage.

MistakeNot4892 replied:
Nvm didn't parse this correctly.

Shadowtail117 committed:
Firedoors will now close on mobs

Add warning sound
Only triggers if people are standing on it
Sanity checks
Reduce damage from 50 to 10
Move the player to the safe side

Spookerton reviewed (approved):
I'm interested in seeing how this goes.
